### ISSUE TITLE
Segfault due to apparent unaligned movaps instructions generated to initialize instance char array member

### ISSUE THREAD

**Issue by clipcarl** [llvm/llvm-project#86895]
I've come across a C++ issue where instantiating some objects in programs compiled by Clang segfaults.

Debugging suggests that Clang is generating SSE `movaps` instructions to initialize some char arrays and it's these instructions that appear to cause segfaults in some circumstances.  I see on the issue tracker that somewhat similar issues with unaligned movaps instructions causing segfaults have been reported many times over the years but it doesn't look like this particular version has been reported or fixed yet.

I've tested on multiple x86-64 Linux systems with the GNU binutils linker with both Clang 16 and Clang 17 with the same results. The problem also occurs when using LLVM's own LLD linker. I don't know if the same problem occurs under other x86-64 operating systems or with other linkers. The problem does **not** occur when using versions of the GCC compiler instead of Clang.

The segfaults occur for some objects under the following minimum set of conditions:
- x86-64 code generation
- Optimization is enabled (-O1 or above)
- Binary built without Clang's default startup code and without the standard C library (`-nstdlib` option to Clang) and of course also without the standard C++ library. (Use of the `-fno-builtin` option makes no difference. I provide my own well-tested memcpy(), memset(), etc. when compiling with `-fno-builtin`.)

Here is minimal code to reproduce. Compile with the following on an x86-64 Linux system:
```
$ clang -O1 -nostdlib -fno-stack-protector -static clang_segv.s clang_segv.cc -o clang_segv
```

clang_segv.cc:
```
struct SegV
{
    void set(const char *s) { char *b = buf; while ( *s ) { *b++ = *s++; } *b = '\0'; }
    char  buf[128] = "";
    char *cursor   = buf; // needed for segfault
};

int
main()
{
    SegV v;
    v.set("aa");
    return 0;
}
```

clang_segv.s:
```
.intel_syntax noprefix

.global _start
_start:
    xor rbp,rbp             # Zero stack base pointer

    xor r9,r9
    pop rdi                 # Pop argc off stack -> rdi for 1st arg to main()
    mov rsi,rsp             # Argv @top of stack -> rsi for 2nd arg to main()
    call main               # Call main()... return result ends up in rax

    xor r9,r9
    mov rdi,rax             # Move main()'s return to 1st argument for exit()
    mov rax,231             # exit_group() syscall
    syscall                 # Tell kernel to exit program
```

This example is the minimum reproducer I could come up with and has no resemblance to the original code where I noticed the problem save that both have objects with char arrays. Changing the code around can mask or unmask the issue which usually suggests a coding error but that does not appear to be the case here.

My debugger seems to think that the problems are the instructions that Clang is generating to initialize the 'buf' char array.
Here is the code the debugger says that Clang generates for the initialization of `char buf[128]`:
```
0x400171 xorps  %xmm0,%xmm0
0x400174 movaps %xmm0,-0x10(%rsp)
0x400179 movaps %xmm0,-0x20(%rsp)
0x40017e movaps %xmm0,-0x30(%rsp)
0x400183 movaps %xmm0,-0x40(%rsp)
0x400188 movaps %xmm0,-0x50(%rsp)
0x40018d movaps %xmm0,-0x60(%rsp)
0x400192 movaps %xmm0,-0x70(%rsp)
0x400197 movaps %xmm0,-0x80(%rsp)
0x40019c lea    -0x80(%rsp),%rax
```
The segfault is apparently generated by the first movaps instruction.

Obviously, I expect that the initialization of the array shouldn't segfault.

It doesn't matter if in-class initialization is used as I do here in this example or if initializer-list initialization is used. Both methods suffer from the same problem. It also doesn't matter if the initialization is done with a different construct such as `char buf[128] = {0};` (which fixes a different Clang bug).

I believe this is an alignment issue with the movaps instructions. If I add alignas(32) to the structure or to the char array the problem goes away. I don't know why I'd need to align at 32 bytes. Perplexingly (to me), aligning at 16 bytes does not mask the problem.

The problem also goes away if I just tell Clang not to generate any SSE instructions with `-mno-sse` but I'd rather not lose those optimizations. In this case Clang uses `movq` instructions to initialize the array instead of `movaps`.

The problem also goes away if I forgo initialization of the char array member and do it manually in the constructor but of course that is less efficient.

Thanks for your help! Please let me know if there's any other information I can provide.

**Comment by efriedma-quic** [comment]
Are you sure your startup code is correct?  The x86-64 ABI requires functions calls to preserve 16-byte stack alignment.

alignas(32) forces clang to emit code to explicitly realign the stack, so it would hide any issue with stack alignment.

**Comment by clipcarl** [comment]
@efriedma-quic 
> Are you sure your startup code is correct? The x86-64 ABI requires functions calls to preserve 16-byte stack alignment.

No, I'm not. I'll look into that shortly and post back here. Thanks and my apologies if it ends up being something simple like that.

**Comment by clipcarl** [comment]
@efriedma-quic - Your hunch was 100% correct. In _start the stack is _already_ aligned so by popping 8 bytes off I was actually *misaligning* it before the call to main(). Even if I'd known this I probably wouldn't have thought it was a big deal because I'm not linking with glibc but you are also correct that the ABI requires it regardless. And because the ABI requires it is perfectly reasonable for Clang to assume that the stack is aligned a certain way upon entry to main() and to generate code accordingly.

Thank you very much for pointing me in the right direction. Sorry for any wasted cycles.

In other words: **This is not a bug**.

**Comment by clipcarl** [comment]
Not a bug.
